yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -207,6 +211,17 @@ private[yarn] class YarnAllocator(
   }
 
   /**
+   * Gets the executor loss reason for a disconnected executor.
+   * Note that this method is expected to be called exactly once per executor ID.
+   */
+  def getExecutorLossReason(executorId: String): ExecutorLossReason = synchronized {
+    allocateResources()
+    // Expect to be asked for a loss reason once and exactly once.
+    assert(completedExecutorExitReasons.contains(executorId))
```

Comment:
By the way I'm open to removing this assertion and building up the map without removing items from it, but I'm not sure of what the memory implications are.